### PR TITLE
Feat/login auth

### PR DESCRIPTION
**Pull request by klavenjones**
Added support for login using the Passport Local strategy. This PR relates to this [ticket](https://trello.com/c/fxMLtKYl)

- Added Passport and used the local passport strategy to aid in user authorization
- Created Login service, modules, and controller 
- Add getUserbyUsername to the user.service, to help retrieve user.

### CODE REVIEW

File: src/api/auth/login/login.controller.spec.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    }).compile();
+
+    controller = module.get<LoginController>(LoginController);
+  });
+
+  it('should be defined', () => {
+    expect(controller).toBeDefined();
+  });
+
+  describe('login', () => {
```
Comment (kaitwalla) [inline src/api/auth/login/login.controller.spec.ts:33]: Needs failed logins as well

File: src/api/auth/login/login.service.spec.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    userService = module.get<UserService>(UserService);
+    loginService = module.get<LoginService>(LoginService);
+  });
+
+  it('should be defined', () => {
+    expect(loginService).toBeDefined();
+  });
+
+  describe('validateUser', () => {
+    it('should call userService.getUserByUsername atleast once.', async () => {
```
Comment (kaitwalla) [inline src/api/auth/login/login.service.spec.ts]: What is this actually testing? Your other test names describe what is happening ("should return null when password does not match"), this is just describing the expect using the same terms

File: src/api/user/user.service.spec.ts [hunk 3]
```diff
@@ -68,15 +69,27 @@ describe('UserService', () => {
...
     });
 
     it('should call userRepository.findByOne with the correct parameters', async () => {
-      await service.getUser(1);
+      await service.getUserById(1);
       expect(userRepository.findOneBy).toHaveBeenCalledWith({ id: 1 });
     });
   });
+
+  describe('getUserByUsername', () => {
```
Comment (kaitwalla) [inline src/api/user/user.service.spec.ts:85]: You did a good job with these. These are hard tests to write because so much of this is abstracted by the library code. If I were writing these from scratch, I might also consider testing the outcomes (findByOne returns user when proper params are passed; returns null or w/e when not). I don't know that it's required here because of the library code, but for me it would bring peace of mind because if the outcome changes, your automated tests would catch it.

Comment (kaitwalla) [reply src/api/user/user.service.spec.ts:85]: Also, from a TDD perspective, outcomes are more useful than calledWIth in this scenario because it could use the proper params but still not provide proper functionality.

**Review by kaitwalla — CHANGES_REQUESTED** [review]

File: src/api/auth/login/login.service.spec.ts (hunk 2 above)
Comment (klavenjones) [reply src/api/auth/login/login.service.spec.ts]: @kaitwalla I was trying to test if the service was being called correctly inside the controller. I'll change the feedback to clarify, or maybe I should remove it.

File: src/api/auth/login/login.controller.spec.ts (hunk 1 above)
Comment (klavenjones) [reply src/api/auth/login/login.controller.spec.ts:33]: @kaitwalla 
Quick question, For context. I wrote tests for failed logins in the e2e tests in the  login.e2e-spec.ts. I wanted to know if I should also do it at the controller level, Currently, @UseGuards(AuthGuard('local'))  are intercepting. Any bad requests (including when a user attempts to log in without being registered) and sends a {"statusCode": 401, "message": "Unauthorized"}, under the hood.  The request won't even reach the login function. Nest abstracts away a lot of Error handling in its library

Should I attempt to mock these guards to test an unauthorized user (There will be more guards down the line)? Or could I rely on the e2e tests to test for proper requests?

Comment (kaitwalla) [reply src/api/auth/login/login.controller.spec.ts:33]: My goal was to unit-test the login controller, but it appears that logic gets tested (in Nest) at the Login service level, which you already did. Nicely done.

**Review by kaitwalla — APPROVED** [review]
